### PR TITLE
fix: add fallback when MBTiles import lacks "name" in metadata

### LINKED ISSUE #95: Importing mbtiles without metadata `name` row fails
If a generated mbtiles file (such as using `tl`) doesn't have a `name` row in the metadata table, importing it fails with a `SQLITE_CONSTRAINT_NOTNULL` error. The offending query is https://github.com/digidem/mapeo-map-server/blob/9457c024f518bcab0fb85a0feba4ed240c7655e1/src/lib/mbtiles_import_worker.js#L94 where we expect a `name` to be present in the extracted mbtiles metadata.

It doesn't seem that the [mbtiles spec](https://github.com/mapbox/mbtiles-spec/blob/master/1.3/spec.md) has any mention of the `name` row being standardized, so we should properly account for it being missing. Think the most proper solution would be updating the [`OfflineArea` table schema](https://github.com/digidem/mapeo-map-server/blob/9457c024f518bcab0fb85a0feba4ed240c7655e1/prisma/schema.prisma#L89) such that it allows the `name` field to be nullable.

### PR DESCRIPTION
**Pull request by EvanHahn**
_I recommend [reviewing this with whitespace changes disabled](https://github.com/digidem/mapeo-map-server/pull/113/files?w=1)._

Before this change, we parsed MBTiles using `any`:

https://github.com/digidem/mapeo-map-server/blob/28a11ea9e37f73b6d9b2f850163cff6c8af25263/src/lib/mbtiles.ts#L83-L84

This meant that (1) required fields could go missing (2) fields could have invalid data. This fixes those problems by avoiding `any`.

It also adds a fallback for `name` in case it's missing, to fix #95.

### CODE REVIEW

File: src/lib/mbtiles.ts [hunk 1]
```diff
@@ -46,7 +51,10 @@ export interface Metadata {
    * should be a number, but node-mbtiles implements this as a string, which
    * is the same as TileJSON */
   version?: string
-  json?: string
```
Comment (EvanHahn) [inline src/lib/mbtiles.ts:49]: We never actually store this field so I removed it.

File: src/lib/mbtiles.ts [hunk 2]
```diff
@@ -77,49 +88,120 @@ export function mbTilesToTileJSON(mbTilesDb: DatabaseInstance): TileJSON {
...
+      )?.vector_layers
+      return Array.isArray(rawVectorLayers) &&
+        rawVectorLayers.every((layer) => validateVectorLayerSchema(layer))
+        ? rawVectorLayers
+        : undefined
+    })(),
+
+    // TODO: Extracted from reference implementation but not sure if it applies for us
+    // https://github.com/mapbox/node-mbtiles/blob/03220bc2fade2ba197ea2bab9cc44033f3a0b37e/lib/mbtiles.js#L300
+    scheme: 'xyz',
```
Comment (achou11) [inline src/lib/mbtiles.ts:164]: if we're not using this field anywhere else in the library, I think it's okay completely omit this. seems like it was an artifact of https://github.com/mapbox/node-mbtiles/issues/65 but I don't think it really applies to us in terms of what this function does (which is just reading the metadata).

File: src/lib/mbtiles.ts [hunk 3]
```diff
@@ -46,7 +51,10 @@ export interface Metadata {
    * should be a number, but node-mbtiles implements this as a string, which
    * is the same as TileJSON */
   version?: string
-  json?: string
+  /** Vector layers describe layers of vector tile data */
+  vector_layers?: VectorLayer[]
+  /** Nonstandard tiling scheme added in reference implementation */
+  scheme?: 'xyz' | 'tms'
```
Comment (achou11) [inline src/lib/mbtiles.ts:57]: see my comment about the scheme field. probably can remove this

**Review by achou11 — APPROVED** [review]
Couple of notes that probably would benefit from getting your thoughts on or confirmation, but otherwise looks great! Appreciate the refactor to make the metadata extraction cleaner and safer

File: src/lib/mbtiles.ts [hunk 4]
```diff
@@ -77,49 +88,120 @@ export function mbTilesToTileJSON(mbTilesDb: DatabaseInstance): TileJSON {
...
-          metadata[row.name] = parseInt(row.value, 10)
-          break
-        case 'center':
-        case 'bounds':
-          metadata[row.name] = row.value.split(',').map(parseFloat)
-          break
-        default:
-          metadata[row.name] = row.value
-          break
+    .reduce((result, { name, value }: { name: unknown; value: unknown }) => {
```
[the remark this answers is not in the document]
Comment (EvanHahn) [reply src/lib/mbtiles.ts:109]: Good point. Changed to `forEach` in 277564251bbbb58b10b2162fcb9c2a259026bfb2.

File: src/lib/mbtiles.ts (hunk 3 above)
Comment (EvanHahn) [reply src/lib/mbtiles.ts]: Dropped in 46053e08923c513c63efc62348aedee84199b039.

File: src/lib/mbtiles.ts [hunk 5]
```diff
@@ -77,49 +88,120 @@ export function mbTilesToTileJSON(mbTilesDb: DatabaseInstance): TileJSON {
...
+
+    vector_layers: (() => {
+      const rawVectorLayers = parseJsonObject(
+        rawMetadata.get('json')
+      )?.vector_layers
+      return Array.isArray(rawVectorLayers) &&
+        rawVectorLayers.every((layer) => validateVectorLayerSchema(layer))
+        ? rawVectorLayers
+        : undefined
+    })(),
```
[the remark this answers is not in the document]
Comment (EvanHahn) [reply src/lib/mbtiles.ts]: Done in da2aaeaab5feac3ae400b813380ca2494dbf8f5d.

File: src/lib/tilejson.ts [hunk 6]
```diff
@@ -36,6 +36,10 @@ const VectorLayerSchema = T.Object({
   source_name: T.Optional(T.String()),
 })
 
+export type VectorLayer = Static<typeof VectorLayerSchema>;
+
+export const validateVectorLayerSchema = ajv.compile(VectorLayerSchema)
```
[the remark this answers is not in the document]
Comment (EvanHahn) [reply src/lib/tilejson.ts:41]: Done in a1fc2b531dfe1b33b8b336f6aa868ec0f8692c6b.

File: test/e2e/tilesets-import.test.js [hunk 7]
```diff
@@ -157,6 +161,31 @@ test('POST /tilesets/import creates style for created tileset', async (t) => {
...
+  const tilesetGetResponse = await server.inject({
+    method: 'GET',
+    url: `/tilesets/${createdTileset.id}`,
+  })
+
+  t.equal(tilesetGetResponse.statusCode, 200)
+
+  const tileset = tilesetGetResponse.json()
+
+  t.equal(tileset.name, 'missing-name-metadata')
```
[the remark this answers is not in the document]
Comment (EvanHahn) [reply test/e2e/tilesets-import.test.js]: Done in b263e4262557394f54863c3ed9c86761dd319c94.

**Review by achou11 — APPROVED** [review]

**Pull request merged**